internal/metrics/v2/counter_test.go

```diff
@@ -0,0 +1,198 @@
+				LabelNames: []string{"ws_close_code"},
+			},
+		},
+		{
+			name:        "should panic when CounterOpts is empty",
+			opts:        CounterOpts{},
+			shouldPanic: true,
+		},
+	}
+	for _, test := range tests {
+		r := prometheus.NewRegistry()
+		t.Run(test.name, func(t *testing.T) {
+			if test.shouldPanic {
+				require.Panics(t, func() {
+					NewCounter(r, test.opts)
+				})
+			} else {
+				require.IsType(t, &prometheusCounter{}, NewCounter(r, test.opts))
+			}
+		})
+	}
+}
+
+func TestPrometheusCounterAdd(t *testing.T) {
+	type fields struct {
+		registry *prometheus.Registry
+		opts     CounterOpts
+	}
+	type args struct {
+		floatVal float64
+		label    []Label
+	}
+	tests := []struct {
+		name        string
+		fields      fields
+		args        args
+		expect      float64
+		shouldPanic bool
+	}{
+		{
+			name: "should add",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: CounterOpts{
+					Subsystem:  "test_subsystem",
+					Name:       "counter_test_total",
+					Help:       "counter_test help",
+					LabelNames: []string{},
+				},
+			},
+			args: args{
+				floatVal: float64(1),
+				label:    []Label{},
+			},
+			expect: float64(1),
+		},
+		{
+			name: "should add with labels",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: CounterOpts{
+					Subsystem:  "test_subsystem",
+					Name:       "counter_test_total",
+					Help:       "counter_test help",
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				floatVal: float64(20),
+				label:    []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+			expect: float64(20),
+		},
+		{
+			name: "should panic when input is < 0",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: CounterOpts{
+					Subsystem:  "test_subsystem",
+					Name:       "counter_test_total",
+					Help:       "counter_test help",
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				floatVal: float64(-2),
+				label:    []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+			shouldPanic: true,
+		},
+		{
+			name: "should succeed when when input is 0",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: CounterOpts{
+					Subsystem:  "test_subsystem",
+					Name:       "counter_test_total",
+					Help:       "counter_test help",
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				label: []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			c := NewCounter(test.fields.registry, test.fields.opts)
+			if test.shouldPanic {
+				require.Panics(t, func() {
+					c.Add(test.args.floatVal, test.args.label...)
+				})
+			} else {
+				c.Add(test.args.floatVal, test.args.label...)
+				family, err := test.fields.registry.Gather()
+				require.NoError(t, err)
+				require.Equal(t, test.expect, family[0].Metric[0].Counter.GetValue())
```

Comment:
In these tests, can we also assert that `MetricFamily.Name`, `MetricFamily.Help`, & `MetricFamily.Type` was set properly?

Reply:
[d883ee3](https://github.com/Kong/koko/pull/364/commits/d883ee35fd047a928573720a02e7dc9999876fe8)

Reply:
Looks good!

You can use the defined metric type constants that the Prometheus protobuf spec defines btw:
```
require.Equal(t, io_prometheus_client.MetricType_COUNTER, family[0].Type)
```

Reply:
I can change it to this if you'd like
```
require.Equal(t, io_prometheus_client.MetricType_COUNTER.String(), family[0].Type.String())
```

Reply:
It's a bit of a nit, but the [example code](https://github.com/Kong/koko/pull/364/files#r1001067920) I posted above would be the ideal way to do it :)

Reply:
[659dcb4](https://github.com/Kong/koko/pull/364/commits/659dcb4efbb0eeb7d300bd58b22b1ffb2b86b782)